contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -74,50 +75,81 @@ pub fn execute_message(
-    ensure_its_source_address(deps.storage, &cc_id.source_chain, &source_address)?;
+    ensure_is_its_source_address(deps.storage, &cc_id.source_chain, &source_address)?;
 
     match HubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)? {
         HubMessage::SendToHub {
             destination_chain,
             message,
-        } => {
-            let destination_address = load_its_contract(deps.storage, &destination_chain)
-                .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
-
-            for chain in [
-                MessageDirection::From(cc_id.source_chain.clone()),
-                MessageDirection::To(destination_chain.clone()),
-            ] {
-                apply_checks(deps.storage, &message, chain)?;
-            }
+        } => execute_hub_message(deps, cc_id, destination_chain, message),
+        _ => bail!(Error::InvalidMessageType),
+    }
+}
 
-            let destination_payload = HubMessage::ReceiveFromHub {
-                source_chain: cc_id.source_chain.clone(),
-                message: message.clone(),
-            }
-            .abi_encode();
-
-            Ok(send_to_destination(
-                deps.storage,
-                deps.querier,
-                destination_chain.clone(),
-                destination_address,
-                destination_payload,
-            )?
-            .add_event(
-                Event::MessageReceived {
-                    cc_id,
-                    destination_chain,
-                    message,
-                }
-                .into(),
-            ))
+fn execute_hub_message(
+    deps: DepsMut,
+    cc_id: CrossChainId,
+    destination_chain: ChainNameRaw,
+    message: Message,
+) -> Result<Response, Error> {
+    let destination_address = load_its_contract(deps.storage, &destination_chain)
+        .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
+
+    apply_to_hub(
+        deps.storage,
+        &cc_id.source_chain,
+        &destination_chain,
+        &message,
+    )?;
+
+    let destination_payload = HubMessage::ReceiveFromHub {
+        source_chain: cc_id.source_chain.clone(),
+        message: message.clone(),
+    }
+    .abi_encode();
+
+    Ok(send_to_destination(
+        deps.storage,
+        deps.querier,
+        &destination_chain,
+        destination_address,
+        destination_payload,
+    )?
+    .add_event(
+        Event::MessageReceived {
+            cc_id,
+            destination_chain,
+            message,
+        }
+        .into(),
+    ))
+}
+
+fn apply_to_hub(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    message: &Message,
+) -> Result<(), Error> {
+    ensure_chain_not_frozen(storage, source_chain)?;
+    ensure_chain_not_frozen(storage, destination_chain)?;
+
+    match message {
+        Message::InterchainTransfer { amount, .. } => {
+            apply_transfer(amount)?;
```

Comment:
the supply change for destination chain will need to be done after the scaling has been applied so `apply_*` should consume the message and return the transformed message for further processing

Reply:
can we not just pass in the scaling factors or something similar here?